sdk/src/main/java/com/mapbox/maps/MapController.kt

```diff
@@ -226,61 +233,72 @@ internal class MapController : MapPluginProviderDelegate, MapControllable {
+    return pluginRegistry.getPlugin(id)
   }
 
-  fun <T> createPlugin(
+  fun createPlugin(
     mapView: MapView?,
-    clazz: Class<T>,
-    vararg constructorArguments: Pair<Class<*>, Any>
-  ): T? = pluginRegistry.createPlugin(mapView, mapInitOptions, clazz, *constructorArguments)
+    descriptor: Plugin
+  ) = pluginRegistry.createPlugin(mapView, mapInitOptions, descriptor)
 
   fun initializePlugins(
     options: MapInitOptions,
     mapView: MapView? = null,
   ) {
-    for (pluginName in options.plugins) {
+    for (plugin in options.plugins) {
       try {
-        val pluginClass = Class.forName(pluginName)
-        val plugin = if (pluginName == PLUGIN_GESTURE_CLASS_NAME) {
-          val attrs = options.attrs
-          if (attrs != null) {
-            createPlugin(
-              mapView,
-              pluginClass,
-              Pair(Context::class.java, options.context),
-              Pair(AttributeSet::class.java, attrs),
-              Pair(Float::class.java, options.mapOptions.pixelRatio)
-            )
-          } else {
-            createPlugin(
-              mapView,
-              pluginClass,
-              Pair(Context::class.java, options.context),
-              Pair(Float::class.java, options.mapOptions.pixelRatio)
-            )
+        val pluginObject = when (plugin.pluginId) {
+          MAPBOX_GESTURES_PLUGIN -> {
+            val attrs = options.attrs
+            if (attrs != null) {
+              GesturesPluginImpl(options.context, attrs, options.mapOptions.pixelRatio)
+            } else {
+              GesturesPluginImpl(options.context, options.mapOptions.pixelRatio)
+            }
+          }
+          MAPBOX_CAMERA_PLUGIN -> {
+            CameraAnimationsPluginImpl()
+          }
+          MAPBOX_ANNOTATION_PLUGIN -> {
+            AnnotationPluginImpl()
+          }
+          MAPBOX_COMPASS_PLUGIN -> {
+            CompassViewPlugin()
+          }
+          MAPBOX_ATTRIBUTION_PLUGIN -> {
+            AttributionViewPlugin()
+          }
+          MAPBOX_LIFECYCLE_PLUGIN -> {
+            MapboxLifecyclePluginImpl()
+          }
+          MAPBOX_LOCATION_COMPONENT_PLUGIN -> {
+            LocationComponentPluginImpl()
+          }
+          MAPBOX_LOGO_PLUGIN -> {
+            LogoViewPlugin()
+          }
+          MAPBOX_MAP_OVERLAY_PLUGIN -> {
+            MapOverlayPluginImpl()
+          }
+          MAPBOX_SCALEBAR_PLUGIN -> {
+            ScaleBarPluginImpl()
+          }
+          else -> {
+            plugin.pluginInstance ?: throw RuntimeException("Custom non Mapbox plugins must have non-null pluginInstance parameter!")
           }
-        } else {
-          createPlugin(mapView, pluginClass)
         }
-        if (plugin is CameraAnimationsPlugin) {
-          mapboxMap.setCameraAnimationPlugin(plugin)
+        createPlugin(mapView, Plugin.Custom(plugin.pluginId, pluginObject))
+        if (pluginObject is CameraAnimationsPluginImpl) {
```

Comment:
should be pluginObject is CameraAnimationsPlugin?

Reply:
No, we want `MapboxMap` extension functions for our plugins only.

Reply:
thinking again, since we always interacting with interface rather than instances, I think it's a better practise to use interface rather than implementation.

> No, we want MapboxMap extension functions for our plugins only.

I'm not sure if we want to limit this to mapbox only, and remove the flexibility to implement the CameraAnimations separately from the one we developed.

wdyt ? @tobrun

Reply:
We have only one reference to the plugin currently in `MapboxMap`. I know it sounds crazy but 2 animations plugins will not work then.
I believe we could easily add ability to have multiple references to gestures / camera plugins if there would be such need but as for now I'd stick to allowing use `MapboxMap` extenstion functions only for our plugins.

Reply:
moreover it did work that way before the change (we were checking java class name to be exactly our impl).

Reply:
@kiryldz I didn't mean two instances of the same type 😂 I meant if user want to switch Mapbox plugin with their own plugin, for example animations, and plugins that's dependent on the animation plugin.

Reply:
👍 let's ticket it out for further discussion, this behaviour matches the previous behaviour.

Reply:
it does not work as expected now as well, we will handle this issue separately from this PR